streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -194,37 +194,56 @@ public KStreamImpl(KStreamBuilder topology, String name, Set<String> sourceNodes
     }
 
     @Override
-    public KStream<K, V> through(Serde<K> keySerde, Serde<V> valSerde, String topic) {
-        to(keySerde, valSerde, topic);
+    public KStream<K, V> through(Serde<K> keySerde, Serde<V> valSerde, String topic, StreamPartitioner<K, V> partitioner) {
+        to(keySerde, valSerde, topic, partitioner);
 
         return topology.stream(keySerde, valSerde, topic);
     }
 
     @Override
+    public KStream<K, V> through(Serde<K> keySerde, Serde<V> valSerde, String topic) {
+        return through(keySerde, valSerde, topic, null);
+    }
+
+    @Override
+    public KStream<K, V> through(String topic, StreamPartitioner<K, V> partitioner) {
+        return through(null, null, topic, partitioner);
+    }
+
+    @Override
     public KStream<K, V> through(String topic) {
-        return through(null, null, topic);
+        return through(null, null, topic, null);
     }
 
     @Override
     public void to(String topic) {
-        to(null, null, topic);
+        to(null, null, topic, null);
+    }
+
+    @Override
+    public void to(String topic, StreamPartitioner<K, V> partitioner) {
+        to(null, null, topic, partitioner);
     }
 
-    @SuppressWarnings("unchecked")
     @Override
     public void to(Serde<K> keySerde, Serde<V> valSerde, String topic) {
+        to(keySerde, valSerde, topic, null);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void to(Serde<K> keySerde, Serde<V> valSerde, String topic, StreamPartitioner<K, V> partitioner) {
         String name = topology.newName(SINK_NAME);
-        StreamPartitioner<K, V> streamPartitioner = null;
 
         Serializer<K> keySerializer = keySerde == null ? null : keySerde.serializer();
         Serializer<V> valSerializer = keySerde == null ? null : valSerde.serializer();
 
-        if (keySerializer != null && keySerializer instanceof WindowedSerializer) {
+        if (partitioner != null && keySerializer != null && keySerializer instanceof WindowedSerializer) {
             WindowedSerializer<Object> windowedSerializer = (WindowedSerializer<Object>) keySerializer;
-            streamPartitioner = (StreamPartitioner<K, V>) new WindowedStreamPartitioner<Object, V>(windowedSerializer);
+            partitioner = (StreamPartitioner<K, V>) new WindowedStreamPartitioner<Object, V>(windowedSerializer);
```

Comment:
Can't follow... What do you mean by
> WindowedStreamPartitioner class file seems missing

I only renamed `streamPartitioner` to `partitioner`... And the cast is required (and was already there) because `WindowedStreamPartitioner` has generic type `Object` for the key instead of `K`